edx_rest_framework_extensions/permissions.py

```diff
@@ -1,9 +1,64 @@
 """ Permission classes. """
 from rest_framework.permissions import BasePermission
-
+from provider.oauth2.models import AccessToken as DOPAccessToken
+from edx_rest_framework_extensions.utils import jwt_decode_handler
 
 class IsSuperuser(BasePermission):
     """ Allows access only to superusers. """
 
     def has_permission(self, request, view):
         return request.user and request.user.is_superuser
+
+class JWTRestrictedApplicationPermission(BasePermission):
+    """
+    This permission class will inspect a request which contains an
+    OAuth2 access_token. The following business logic is applied:
+    1) Is the OAuth2 token passed in a legacy django-oauth-provider (DOP) token, if so
+       all applications connecting via DOP reflect trusted internal applications
+    2) Is the access_token associated with a RestrictedApplication? If not,
+       the caller is viewed as a trusted application and the permission check is passed
+    3) If the access_token is associated with a RestrictedApplication, then:
+        2a) Get the 'scopes' from the access_token and inspect the passed in 'view'
+        2b) Inspect the view object for a 'required_scopes' attribute on the view
+        2c) If there is no 'required_scopes', then FAIL THE REQUEST, since the
+            view has not declared how to allow RestrictedApplication to access it
+        2d) If there is a 'required_scopes' attribute on the view object then make
+            sure that the access_token has that scope on it
+        2e) If access_token does not contain the 'required_scopes' then fail the request
+        2f) If all above checks succees, pass the permissions check
+    """
+    def _token_filters(self, token):
+        # get filters list from jwt token and return dict
+        if jwt_decode_handler(token).has_key('filters'):
+	    filters_list = jwt_decode_handler(token)['filters']
```

Comment:
The indentation is not right here.

Reply:
Can you pass the decoded token to this function since you already decoded in `has_permission`?

Reply:
fixed the indentation, nice catch. also, passed decoded token